### PR TITLE
Username idx

### PR DESCRIPTION
**Pull request by alpe**
* Add multi key index
* Chain address must not be null

### CODE REVIEW

File: orm/bucket.go [hunk 1]
```diff
@@ -31,6 +31,13 @@ var (
 	isBucketName = regexp.MustCompile(`^[a-z_]{3,10}$`).MatchString
 )
 
+type Indexed interface {
```
Comment (ethanfrey) [inline orm/bucket.go]: Nice, two hold multiple index types

File: orm/bucket.go [hunk 2]
```diff
@@ -203,14 +210,24 @@ func (b Bucket) Sequence(name string) Sequence {
 //
 // Designed to be chained.
 func (b Bucket) WithIndex(name string, indexer Indexer, unique bool) Bucket {
+	return b.WithMultiKeyIndex(name, func(obj Object) ([][]byte, error) {
```
Comment (ethanfrey) [inline orm/bucket.go]: Question: why use Indexed interface then only have one concrete implementation?

I guess you want to generalize this when pr finished?

Comment (ethanfrey) [reply orm/bucket.go]: Then again, MultiIndexer is a superset of Indexer, so you can get away with this

File: orm/bucket.go [hunk 3]
```diff
@@ -203,14 +210,24 @@ func (b Bucket) Sequence(name string) Sequence {
...
+		key, err := indexer(obj)
+		if err != nil {
+			return nil, err
+		}
+		return [][]byte{key}, nil
```
Comment (ethanfrey) [inline orm/bucket.go]:
```
if key == nil {
  return nil, nil
}
```

this is important case, to remove the index (0 or 1 results)

File: orm/index.go [hunk 4]
```diff
@@ -20,17 +23,34 @@ type Index struct {
...
 // Indexer calculcates the index for an object
 // unique enforces a unique constraint on the index
 // refKey calculates the absolute dbkey for a ref
 func NewIndex(name string, indexer Indexer, unique bool,
+	refKey func([]byte) []byte) Index {
+	// TODO: index name must be [a-z_]
+	return Index{
+		name: name,
+		id:   append(indPrefix, []byte(name+":")...),
+		index: func(obj Object) ([][]byte, error) {
```
Comment (ethanfrey) [inline orm/index.go]: seen this closure twice now...

File: orm/bucket.go (hunk 3 above)
Comment (ethanfrey) [reply orm/bucket.go]: Or do I enforce it to be non-nil below? If so we should return error on nil

File: orm/index.go [hunk 5]
```diff
@@ -70,17 +90,27 @@ func (i Index) Update(db weave.KVStore, prev Object, save Object) error {
 	case s{true, true}:
 		return ErrUpdateNil()
 	case s{true, false}:
-		key, err := i.index(save)
+		keys, err := i.index(save)
 		if err != nil {
 			return err
 		}
-		return i.insert(db, key, save.Key())
+		for _, key := range keys {
```
Comment (ethanfrey) [inline orm/index.go]: :+1:

File: orm/index.go [hunk 6]
```diff
@@ -90,11 +120,14 @@ func (i Index) Update(db weave.KVStore, prev Object, save Object) error {
...
 // returns a list of all pk that match (may be empty), or an error
 func (i Index) GetLike(db weave.ReadOnlyKVStore, pattern Object) ([][]byte, error) {
-	index, err := i.index(pattern)
+	indexes, err := i.index(pattern)
 	if err != nil {
 		return nil, err
 	}
-	return i.GetAt(db, index)
+	for _, index := range indexes {
+		return i.GetAt(db, index)
```
Comment (ethanfrey) [inline orm/index.go]: So this returns first match, doesn't concatenate them all?

File: orm/index.go [hunk 7]
```diff
@@ -90,11 +120,14 @@ func (i Index) Update(db weave.KVStore, prev Object, save Object) error {
 // GetLike calculates the index for the given pattern, and
 // returns a list of all pk that match (may be empty), or an error
 func (i Index) GetLike(db weave.ReadOnlyKVStore, pattern Object) ([][]byte, error) {
-	index, err := i.index(pattern)
+	indexes, err := i.index(pattern)
```
Comment (ethanfrey) [inline orm/index.go]: Yeah, I am not sure of this pattern when there is multi-index...
Maybe we can remove it if not heavily used....

File: orm/index.go [hunk 8]
```diff
@@ -183,33 +216,61 @@ func (i Index) move(db weave.KVStore, prev Object, save Object) error {
...
-		k := i.IndexKey(newKey)
-		val := db.Get(k)
-		if val != nil {
-			return ErrUniqueConstraint(i.name)
+	newKeysSet := make(map[string]struct{})
+	for _, v := range newKeys {
+		newKeysSet[string(v)] = struct{}{}
+	}
+
+	// check unique constraints first
```
Comment (ethanfrey) [inline orm/index.go]: maybe calculate `toAddSet` and `toRemoveSet` as diffs, then some of the below code is simplified

File: orm/index.go [hunk 9]
```diff
@@ -183,33 +216,61 @@ func (i Index) move(db weave.KVStore, prev Object, save Object) error {
...
+			if err = i.remove(db, oldKey, prev.Key()); err != nil {
+				return err
+			}
+			continue
+		}
+	}
+
+	// add new keys
+	for _, newKey := range newKeys {
+		// check unique constraint before removing
```
Comment (ethanfrey) [inline orm/index.go]: `s/removing/inserting/`

also above (my typo in original)

File: orm/index.go [hunk 10]
```diff
@@ -183,33 +216,61 @@ func (i Index) move(db weave.KVStore, prev Object, save Object) error {
...
+		newKeysSet[string(v)] = struct{}{}
+	}
+
+	// check unique constraints first
+	for _, newKey := range newKeys {
+		if _, exists := oldKeysSet[string(newKey)]; exists { // skip existing keys
+			continue
+		}
+		// check unique constraint before removing
+		if i.unique && len(oldKeysSet) != 0 {
```
Comment (ethanfrey) [inline orm/index.go]: why is this `len(oldKeysSet) != 0` check there?

maybe this is important? maybe there was some odd logic in my old code?
good to add comment if we really want it

File: x/nft/username/bucket.go [hunk 11]
```diff
@@ -42,3 +46,21 @@ func (b Bucket) Create(db weave.KVStore, owner weave.Address, id []byte, approva
 	}
 	return obj, humanAddress.SetChainAddresses(owner, addresses)
 }
+
+func chainAddressIndexer(obj orm.Object) ([][]byte, error) {
```
Comment (ethanfrey) [inline x/nft/username/bucket.go]: nice function.

File: x/nft/username/bucket.go [hunk 12]
```diff
@@ -16,7 +19,8 @@ type Bucket struct {
 
 func NewBucket() Bucket {
 	return Bucket{
-		Bucket: nft.WithOwnerIndex(orm.NewBucket(BucketName, NewUsernameToken(nil, nil, nil))),
+		Bucket: nft.WithOwnerIndex(orm.NewBucket(BucketName, NewUsernameToken(nil, nil, nil))).
+			WithMultiKeyIndex(ChainAddressIndexName, chainAddressIndexer, true),
```
Comment (ethanfrey) [inline x/nft/username/bucket.go]: cool, global unique address lookup.
:+1:

File: x/nft/username/handler_test.go [hunk 13]
```diff
@@ -42,49 +42,56 @@ func TestHandleIssueTokenMsg(t *testing.T) {
...
-		{ // key already used
+		{ // name already used
 			owner:           alice.Address(),
 			id:              []byte("existing@example.com"),
-			details:         username.TokenDetails{[]username.ChainAddress{myAddress}},
+			details:         username.TokenDetails{myNewChainAddresses},
+			expCheckError:   false,
+			expDeliverError: true,
+		},
+		{ // address already used
```
Comment (ethanfrey) [inline x/nft/username/handler_test.go]: nice test

File: x/nft/username/handler_test.go [hunk 14]
```diff
@@ -184,15 +192,19 @@ func TestQueryTokenByName(t *testing.T) {
...
-	b, _ := blockchains.Create(db, alice.Address(), []byte("myNet"), nil, nil)
-	blockchains.Save(db, b)
-	b, _ = blockchains.Create(db, alice.Address(), []byte("myOtherNet"), nil, nil)
-	blockchains.Save(db, b)
+	for _, blockchainID := range []string{"myNet", "myOtherNet"} {
+		b, _ := blockchains.Create(db, alice.Address(), []byte(blockchainID), nil, nil)
+		blockchains.Save(db, b)
+	}
+
+	myAddress := []username.ChainAddress{{ChainID: []byte("myNet"), Address: []byte("myChainAddress")}}
```
Comment (ethanfrey) [inline x/nft/username/handler_test.go]: look like solid test updates... didn't realize they were here

File: orm/index.go (hunk 10 above)
Comment (alpe) [reply orm/index.go]: I stumbled upon this, too. In the original code an old key must exists to execute the unique check here. I tried to mimic this behaviour with the `len` check. anyway, I am more than happy to remove the extra condition.

File: orm/index.go (hunk 9 above)
Comment (alpe) [reply orm/index.go]: the constraint is checked on the insert function as well. So it makes sense only as you wrote it before to check before removing. Although I assume this is happening within a transaction.

File: orm/index.go [hunk 15]
```diff
@@ -118,16 +123,35 @@ func (i Index) Update(db weave.KVStore, prev Object, save Object) error {
...
+		r = append(r, pks...)
+	}
+	return deduplicate(r), nil
+}
+
+func deduplicate(r [][]byte) [][]byte {
+	for i, v := range r {
+		for j := i + 1; j < len(r); j++ {
+			if bytes.Equal(v, r[j]) {
+				r = append(r[0:j], r[j+1:]...)
```
Comment (ethanfrey) [inline orm/index.go]: is it safe to re-assign r while in the range loop?

code looks good, I would have a separate `res` variable, but I do see the code is smaller here, even if a bit confusing modifying a loop guard in the loop

File: orm/index.go [hunk 16]
```diff
@@ -224,24 +248,12 @@ func (i Index) move(db weave.KVStore, prev Object, save Object) error {
...
-	oldKeysSet := make(map[string]struct{})
-	for _, v := range oldKeys {
-		oldKeysSet[string(v)] = struct{}{}
-	}
-
-	newKeysSet := make(map[string]struct{})
-	for _, v := range newKeys {
-		newKeysSet[string(v)] = struct{}{}
-	}
+	keysToAdd := subtract(newKeys, oldKeys)
```
Comment (ethanfrey) [inline orm/index.go]: :)

File: orm/index.go (hunk 15 above)
Comment (alpe) [reply orm/index.go]: I remembered it was safe unlike java but when I searched for some resources on the net I am not sure anymore. It definitively was not save with older versions of Go in <=2017. Anyway, I refactored into a classic for loop to be on the save side. 
fyi: I found this wiki entry which has a nice trick to reuse memory with append: https://github.com/golang/go/wiki/SliceTricks#filtering-without-allocating

File: orm/index_test.go [hunk 17]
```diff
@@ -115,6 +115,174 @@ func TestCounterIndex(t *testing.T) {
...
+			for _, k := range spec.expNotKeys {
+				pks, err := idx.GetAt(db, k)
+				require.NoError(t, err)
+				assert.Nil(t, pks)
+			}
+		})
+	}
+}
+
+func TestGetLikeWithMultiKeyIndex(t *testing.T) {
```
Comment (ethanfrey) [inline orm/index_test.go]: Makes sense.
Matches if any of the indexes match.

Thanks for the test

File: x/nft/username/handler.go [hunk 18]
```diff
@@ -73,7 +76,7 @@ func (h IssueHandler) Deliver(ctx weave.Context, store weave.KVStore, tx weave.T
 	if err != nil {
 		return res, err
 	}
-
+	res.Tags = append(res.Tags, common.KVPair{Key: []byte(msgTypeTagKey), Value: []byte(newUsernameTagValue)})
```
Comment (ethanfrey) [inline x/nft/username/handler.go]: I like this. We should do this for all messages.
Maybe we could auto-perform this with a middleware to tag `msgType` to the `msg.Path()`?
That would be a separate PR in any case.

File: x/nft/username/handler_test.go [hunk 19]
```diff
@@ -148,6 +157,29 @@ func TestHandleIssueTokenMsg(t *testing.T) {
...
+	// when
+	tx := helpers.MockTx(&username.IssueTokenMsg{
+		Owner:   alice.Address(),
+		Id:      []byte("any@example.com"),
+		Details: username.TokenDetails{[]username.ChainAddress{{ChainID: []byte("myNet"), Address: []byte("myChainAddress")}}},
+	})
+	res, err := handler.Deliver(nil, db, tx)
+	// then
+	require.NoError(t, err)
+	assert.Equal(t, []common.KVPair{{[]byte("msgType"), []byte("registerUsername")}}, res.Tags)
```
Comment (ethanfrey) [inline x/nft/username/handler_test.go]: As mentioned above, I like.
But maybe a separate PR (different topic, right?) and we can seek to generalize this approach

File: x/nft/username/bucket.go (hunk 12 above)
Comment (ethanfrey) [reply x/nft/username/bucket.go]: Is this tested somewhere? Probably it is, I just missed it, please link to lines so I can review

File: x/nft/username/handler.go (hunk 18 above)
Comment (alpe) [reply x/nft/username/handler.go]: A middleware makes sense when want to go that route. At this stage we have a specific use case we want to cover. I am afraid we could end up with tons of unused tags. Let's think about some other use cases we want to handle first.

Comment (ethanfrey) [reply x/nft/username/handler.go]: :+1:

**Comment by alpe** [comment]
@ethanfrey thanks for the feedback. I have applied the changes. Tests for queries are added and setting the tag is moved to #158
